Implement dialog.requestClose()

lukewarlow opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

lukewarlow committed:
Implement dialog.requestClose()

lukewarlow commented:
Old PR that I forgot to push before. Unfortunately the WPTs depend on AbortController which Servo doesn't have.

simonwuelker commented:
I would expect at least some idlharness tests to pass

simonwuelker commented on components/script/dom/htmldialogelement.rs:
```diff
@@ -124,4 +125,25 @@ impl HTMLDialogElementMethods<crate::DomTypeHolder> for HTMLDialogElement {
             .queue_simple_event(target, atom!("close"));
     }
+
+    // https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-requestclose
```
```diff
- // https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-requestclose
+ /// <https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-requestclose>

simonwuelker commented on components/script/dom/htmldialogelement.rs:
```diff
@@ -124,4 +125,25 @@ impl HTMLDialogElementMethods<crate::DomTypeHolder> for HTMLDialogElement {
+
+    // https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-requestclose
+    fn RequestClose(&self, return_value: Option<DOMString>) {
+        let element = self.upcast::<Element>();
+        let target = self.upcast::<EventTarget>();
+
+        if !element
+            .has_attribute(&local_name!("open"))
+        {
+            return;
```
Can you add spec comments for this function?

Something like
```
// Step 1. If thisdoes not have an open attribute, then return.
```
 with TODOs for missing steps

simonwuelker commented on components/script/dom/htmldialogelement.rs:
```diff
@@ -124,4 +125,25 @@ impl HTMLDialogElementMethods<crate::DomTypeHolder> for HTMLDialogElement {
+
+        let window = self.owner_window();
+        let event = Event::new(window.upcast(), atom!("cancel"), EventBubbles::DoesNotBubble, EventCancelable::Cancelable, CanGc::note());
+        let event_status = event.fire(target, CanGc::note());
```
if you add a `canGc` entry for this method in `Bindings.conf` you'll get a `CanGc` argument that you can use here

simonwuelker reviewed (commented):
I'm a bit confused, the implementation of `RequestClose` seems to do very different things than the spec? The spec doesn't talk about a `"close"` event at all.

lukewarlow commented:
The spec is complicated because of the integration with dialogs close watcher infrastructure. This isn't implemented in servo though and the method can be simplified to what I've implemented (it follows the same flow as what I implemented in WebKit too).

lukewarlow committed:
Address comments

- Make CanGC change
- Add spec comments
- Update idlharness expectations

gterzian commented on components/script/dom/htmldialogelement.rs:
```diff
@@ -124,4 +125,36 @@ impl HTMLDialogElementMethods<crate::DomTypeHolder> for HTMLDialogElement {
+        // 5. Set this's request close return value to returnValue.
+        // 6. Request to close dialog's close watcher with false.
+        // 7. Set dialog's enable close watcher for requestClose() to false.
+        // ADHOC: This implementation doesn't follow the steps above, but has the same observable behavior.
```
In general I think not following the spec requires a good reason(which I don't see here), but in this particular case, given the explanation for [`close-watcher-manager`](https://html.spec.whatwg.org/multipage/interaction.html#close-watcher-manager) is "anti-abuse protections designed to prevent developers from disabling users' history traversal abilities", I am against implementing this method without the proper backing infrastructure.

lukewarlow replied:
This method would "bypass" that anti abuse mechanism per the spec. So this doesn't introduce anything bad in that way. That anti abuse mechanism is designed for Android back gestures as close Requests to close watchers.

This is a JS method and so doesn't have the same issue of being used for multiple things.

lukewarlow replied:
To be specific:

Step 6: "[Request to close](https://html.spec.whatwg.org/multipage/interaction.html#close-watcher-request-close) dialog's [close watcher](https://html.spec.whatwg.org/multipage/interactive-elements.html#dialog-close-watcher) with false." - the false here, goes through and is used in step 6 of the [close watch request to close algorithm](https://html.spec.whatwg.org/multipage/interaction.html#close-watcher-request-close).

Where the false value bypasses the history-action user activation checks (the core of the anti abuse mechanics).

lukewarlow replied:
Having said that I did also get some draft PR for close watchers but ran into some weird issues regarding Pointer types so perhaps I can push that up and see if we can work it out. Just thought I'd point out that this isn't an actually bad bypass to anything.

gterzian reviewed (changes requested):
Thanks for looking into this. Given the reason for the close watcher infrastructure, I think implementing it is a prerequisite for implementing this method.
